src/libraries/System.Text.Json/tests/Serialization/ReferenceHandlerTests.IgnoreCycle.cs

```diff
@@ -0,0 +1,324 @@
+        {
+            var root = new WrapperForIDictionary();
+            root.Add("self", root);
+            await Test_Serialize_And_SerializeAsync(root, @"{""self"":null}", s_optionsIgnoreCycles);
+        }
+
+        [Fact]
+        public async Task IgnoreCycles_OnArray()
+        {
+            var root = new object[1];
+            root[0] = root;
+            await Test_Serialize_And_SerializeAsync(root, "[null]", s_optionsIgnoreCycles);
+        }
+
+        [Theory]
+        [InlineData(typeof(List<object>))]
+        [InlineData(typeof(GenericIListWrapper<object>))]
+        public async Task IgnoreCycles_OnLists(Type typeToSerialize)
+        {
+            var root = (IList<object>)Activator.CreateInstance(typeToSerialize);
+            root.Add(root);
+            await Test_Serialize_And_SerializeAsync(root, "[null]", s_optionsIgnoreCycles);
+        }
+
+        [Theory]
+        [InlineData(typeof(GenericISetWrapper<object>))]
+        [InlineData(typeof(GenericICollectionWrapper<object>))]
+        public async Task IgnoreCycles_OnCollections(Type typeToSerialize)
+        {
+            var root = (ICollection<object>)Activator.CreateInstance(typeToSerialize);
+            root.Add(root);
+            await Test_Serialize_And_SerializeAsync(root, "[null]", s_optionsIgnoreCycles);
+        }
+
+        [Fact]
+        public async Task IgnoreCycles_OnCollections_WithoutAddMethod()
+        {
+            var root = new Stack<object>();
+            root.Push(root);
+            await Test_Serialize_And_SerializeAsync(root, "[null]", s_optionsIgnoreCycles);
+
+            var root2 = new Queue<object>();
+            root2.Enqueue(root2);
+            await Test_Serialize_And_SerializeAsync(root2, "[null]", s_optionsIgnoreCycles);
+
+            var root3 = new ConcurrentStack<object>();
+            root3.Push(root3);
+            await Test_Serialize_And_SerializeAsync(root3, "[null]", s_optionsIgnoreCycles);
+
+            var root4 = new ConcurrentQueue<object>();
+            root4.Enqueue(root4);
+            await Test_Serialize_And_SerializeAsync(root4, "[null]", s_optionsIgnoreCycles);
+
+            var root5 = new Stack();
+            root5.Push(root5);
+            await Test_Serialize_And_SerializeAsync(root5, "[null]", s_optionsIgnoreCycles);
+
+            var root6 = new Queue();
+            root6.Enqueue(root6);
+            await Test_Serialize_And_SerializeAsync(root6, "[null]", s_optionsIgnoreCycles);
+        }
+
+        [Fact]
+        public async Task IgnoreCycles_OnExtensionData()
+        {
+            var root = new EmptyClassWithExtensionProperty();
+            root.MyOverflow.Add("root", root);
+            await Test_Serialize_And_SerializeAsync(root, @"{""root"":null}", s_optionsIgnoreCycles);
+        }
+
+        [Fact]
+        public async Task IgnoreCycles_OnBoxedValueTypes()
+        {
+            IValueNode root = new ValueNode();
```

Comment:
Nice that this uses an interface. Can we also have a test where `object` is the declared type?